src/mbgl/style/sources/raster_dem_source.cpp

```diff
@@ -1,18 +1,77 @@
+#include <mbgl/storage/file_source.hpp>
 #include <mbgl/style/conversion/json.hpp>
 #include <mbgl/style/conversion/tileset.hpp>
 #include <mbgl/style/layer.hpp>
 #include <mbgl/style/source_observer.hpp>
 #include <mbgl/style/sources/raster_dem_source.hpp>
 #include <mbgl/style/sources/raster_source_impl.hpp>
 #include <mbgl/tile/tile.hpp>
+#include <mbgl/util/async_request.hpp>
+#include <mbgl/util/exception.hpp>
 #include <mbgl/util/mapbox.hpp>
 #include <utility>
 
 namespace mbgl {
 namespace style {
 
-RasterDEMSource::RasterDEMSource(std::string id, variant<std::string, Tileset> urlOrTileset_, uint16_t tileSize)
-    : RasterSource(std::move(id), std::move(urlOrTileset_), tileSize, SourceType::RasterDEM) {}
+RasterDEMSource::RasterDEMSource(std::string id,
+                                 variant<std::string, Tileset> urlOrTileset_,
+                                 uint16_t tileSize,
+                                 std::optional<Tileset::DEMEncoding> encoding)
+    : RasterSource(std::move(id), std::move(urlOrTileset_), tileSize, SourceType::RasterDEM),
+      encoding(encoding) {}
+
+void RasterDEMSource::loadDescription(FileSource& fileSource) {
+    if (urlOrTileset.is<Tileset>()) {
+        baseImpl = makeMutable<Impl>(impl(), urlOrTileset.get<Tileset>());
+        loaded = true;
+        observer->onSourceLoaded(*this);
+        return;
+    }
+
+    if (req) {
+        return;
+    }
+
+    const auto& rawURL = urlOrTileset.get<std::string>();
+    const auto& url = util::mapbox::canonicalizeSourceURL(fileSource.getResourceOptions().tileServerOptions(), rawURL);
+
+    req = fileSource.request(Resource::source(url), [this, url, &fileSource](const Response& res) {
+        if (res.error) {
+            observer->onSourceError(*this, std::make_exception_ptr(std::runtime_error(res.error->message)));
+        } else if (res.notModified) {
+            return;
+        } else if (res.noContent) {
+            observer->onSourceError(*this, std::make_exception_ptr(std::runtime_error("unexpectedly empty TileJSON")));
+        } else {
+            conversion::Error error;
+            std::optional<Tileset> tileset = conversion::convertJSON<Tileset>(*res.data, error);
+            if (!tileset) {
+                observer->onSourceError(*this, std::make_exception_ptr(util::StyleParseException(error.message)));
+                return;
+            }
+            const auto& tileServerOptions = fileSource.getResourceOptions().tileServerOptions();
+            if (tileServerOptions.uriSchemeAlias() == "mapbox") {
+                util::mapbox::canonicalizeTileset(tileServerOptions, *tileset, url, getType(), getTileSize());
+            }
+
+            if (encoding) {
```

Comment:
@louwers This is the difference. We override the encoding if it's different. This is the same approach employed by [VectorSource::loadDescription](https://github.com/maplibre/maplibre-native/blob/ec531339266acf2b8439ab377e6bb77e8ccc849c/src/mbgl/style/sources/vector_source.cpp#L44) which is virtually identical other than overriding the min/max zoom levels.

This duplication of logic across the different tile-based sources is the problem that I tried to address in the other proposal through an extraction of a common super class to hold the shared `loadDescription`.

Reply:
I did look at that as well. Turns out the logic is also the pretty much the same as `VectorSource` (or any TileJSON-based source) which is how I got to #3570.